rex/bias_correction.py

```diff
@@ -0,0 +1,578 @@
+        self.sampling = str(self._clean_kwarg(sampling))
+        self.log_base = float(self._clean_kwarg(log_base))
+        self.scipy_dist = None
+
+        if self.dist_name != 'empirical':
+            self.scipy_dist = getattr(scipy.stats, self.dist_name, None)
+            if self.scipy_dist is None:
+                msg = ('Could not get requested distribution "{}" from '
+                       '``scipy.stats``. Please double check your spelling '
+                       'and select "empirical" or one of the continuous '
+                       'distribution options from here: '
+                       'https://docs.scipy.org/doc/scipy/reference/stats.html'
+                       .format(self.dist_name))
+                logger.error(msg)
+                raise KeyError(msg)
+
+    @staticmethod
+    def _clean_kwarg(inp):
+        """Clean any kwargs inputs (e.g., dist, relative) that might be
+        provided as an array and must be collapsed into a single string or
+        boolean value"""
+        unique = np.unique(inp)
+        msg = ('_QuantileDeltaMapping kwargs must have only one unique input '
+               'even if being called with arrays as part of reV but found: {}'
+               .format(unique))
+        assert len(unique) == 1, msg
+
+        while isinstance(inp, np.ndarray):
+            inp = inp[0]
+
+        return inp
+
+    def _clean_params(self, params, arr_shape):
+        """Verify and clean 2D parameter arrays for passing into empirical
+        distribution or scipy continuous distribution functions.
+
+        Parameters
+        ----------
+        params : np.ndarray
+            Input params shape should be (space, N) where N is the number of
+            parameters for the distribution.
+        arr_shape : tuple
+            Array shape should be (time, space).
+
+        Returns
+        -------
+        params : np.ndarray | list
+            If a scipy continuous dist is set, this output will be params
+            unpacked along axis=1 into a list so that the list entries
+            represent the scipy distribution parameters
+            (e.g., shape, scale, loc) and each list entry is of shape (space,)
+        """
+
+        msg = f'params must be 2D array but received {type(params)}'
+        assert isinstance(params, np.ndarray), msg
+
+        if len(params.shape) == 1:
+            params = np.expand_dims(params, 0)
+
+        msg = (f'params must be 2D array of shape ({arr_shape[1]}, N) '
+               f'but received shape {params.shape}')
+        assert len(params.shape) == 2, msg
+        assert params.shape[0] == arr_shape[1], msg
+
+        if self.scipy_dist is not None:
+            params = [params[:, i] for i in range(params.shape[1])]
+
+        return params
+
+    def _get_quantiles(self, n_samples):
+        """If dist='empirical', this will get the quantile values for the CDF
+        x-values specified in the input params"""
+
+        if self.sampling == 'linear':
```

Comment:
This looks similar to code in the `cdf` function (from `temporal_stats.py`). Could consider pulling it out into it's own small utility function and calling that in both places?

Reply:
moved to the new actual bc_utils.py file